source/server/guarddog_impl.cc

```diff
@@ -111,46 +152,29 @@ void GuardDogImpl::stopWatching(WatchDogSharedPtr wd) {
   }
 }
 
-bool GuardDogImpl::waitOrDetectStop() {
-  force_checked_event_.notifyAll();
-  Thread::LockGuard guard(exit_lock_);
-  // Spurious wakeups are OK without explicit handling. We'll just check
-  // earlier than strictly required for that round.
-
-  // Preferably, we should be calling
-  //   time_system_.waitFor(exit_lock_, exit_event_, loop_interval_);
-  // here, but that makes GuardDogMissTest.* very flaky. The reason that
-  // directly calling condvar waitFor works is that it doesn't advance
-  // simulated time, which the test is carefully controlling.
-  //
-  // One alternative approach that would be easier to test is to use a private
-  // dispatcher and a TimerCB to execute the loop body of threadRoutine(). In
-  // this manner, the same dynamics would occur in production, with added
-  // overhead from libevent, But then the unit-test would purely control the
-  // advancement of time, and thus be more robust. Another variation would be
-  // to run this watchdog on the main-thread dispatcher, though such an approach
-  // could not detect when the main-thread was stuck.
-  exit_event_.waitFor(exit_lock_, loop_interval_); // NO_CHECK_FORMAT(real_time)
```

Comment:
It's probably worth pointing out here that this condvar-wait is entirely replaced by libevent in the new impl, via the use of a new private dispatcher owned by the guard-dog. We *still* need a condvar to avoid spinning over dispatcher-run while waiting for loop_interval_, but it's an untimed condvar wait, signaled by the scheduling of a timer.

It occurs to me that I could alternatively implement the Event::Scheduler interface with a new class CondvarScheduler which would the boil down to the exact same system-call we have here in the current version of guarddog_impl.cc. The benefit of this would be that we'd avoid carrying a dispatcher and a libevent base-ptr in the guarddog when there are no network or file-events to watch, simplifying the logic in operation. The cost is that implementing CondvarScheduler in general would add complexity to the PR, probably refactoring some of the SimulatedTimeSystem implementation of the alarm-set into something that would be used at runtime. As this functionality is already in libevent it seems simpler to re-use that.